lib/HLSL/HLOperationLower.cpp

```diff
@@ -3607,12 +3611,36 @@ void TranslateLoad(ResLoadHelper &helper, HLResource::Kind RK,
   }
 
   if (DXIL::IsStructuredBuffer(RK)) {
-    // Basic type case for StructuredBuffer::Load()
-    Value *ResultElts[4];
-    Value *StructBufLoad = GenerateStructBufLd(helper.handle, helper.addr, OP->GetU32Const(0),
-      helper.status, EltTy, ResultElts, OP, Builder, numComponents, Alignment);
-    dxilutil::MigrateDebugValue(helper.retVal, StructBufLoad);
-    Value *retValNew = ScalarizeElements(Ty, ResultElts, Builder);
+    Type* RetTy = helper.retVal->getType();
+    unsigned ElemCount = RetTy->isVectorTy() ? RetTy->getVectorNumElements() : 1;
+    std::vector<Value*> bufLds;
+    const bool isBool = EltTy->isIntegerTy(1);
+
+    // Bool are represented as i32 in memory
+    Type* MemReprTy = isBool ? Builder.getInt32Ty() : EltTy;
+    Value* retValNew = TranslateStructBufVecLd(MemReprTy, ElemCount, Builder, helper.handle, OP, helper.status,
+      helper.addr, OP->GetU32Const(0), DL, bufLds);
+    DXASSERT_NOMSG(!bufLds.empty());
+    dxilutil::MigrateDebugValue(helper.retVal, bufLds.front());
+
+    if (isBool) {
+      // Convert result back to register representation.
+      retValNew = Builder.CreateICmpNE(retValNew, Constant::getNullValue(retValNew->getType()));
+    }
+
+    // We always get a vector from TranslateStructBufVecLd(). For vec1 case,
+    // convert it to scalar if needed to match the target type.
+    // Ex 1: For stbuf.Load<float1x1>(i), both the new value type and
+    // the target type is <1 x float>. No conversion needed here.
+    //
+    // Ex 2: For stbuf.Load<float>(i), the new value type is <1 x float>, but
+    // the target type is float, so we need a conversion here.
+    Type* NewValTy = retValNew->getType();
+    if (NewValTy->isVectorTy() && NewValTy->getVectorNumElements() == 1 &&
+       helper.retVal->getType() == NewValTy->getVectorElementType()) {
```

Comment:
Is the only way the `helper.retVal->getType()` and `NewValTy->getVectorElementType()` differ is by one being a vector and the other not? If so, is this really just checking if helper.retVal is a scalar?

I don't see how they could be different component types

Reply:
I cleaned it up a bit to make it less confusing. I am now passing a boolean flag to indicate in `TranslateStructBufVecLd` to indicate when building vector can be skipped in the first place.  

`TranslateStructBufMatLd` -> `TranslateStructBufVecLd`  // Can only have vector types 
`TranslateLoad` -> `TranslateStructBufVecLd` // Can have both scalar and vector types